gk/fib.c

```diff
@@ -0,0 +1,1640 @@
+
+	return gt_fib;
+
+del_ip_prefix_fib:
+	lpm_del_route(gt_addr, gt_prefix_len, ltbl);
+
+	gt_fib->u.grantor.next_fib->ref_cnt--;
+	ip_prefix_fib->u.grantor.eth_cache->ref_cnt--;
+	initialize_fib_entry(ip_prefix_fib);
+
+del_gt_fib:
+	gt_fib->u.grantor.next_fib->ref_cnt--;
+	/* TODO Add concurrency control. */
+	eth_cache->ref_cnt--;
+
+	if (is_neighbor && eth_cache->ref_cnt == 0)
+		neigh_del_ether_cache(neigh_ht, &gt_addr->ip);
+
+	initialize_fib_entry(gt_fib);
+	return NULL;
+}
+
+static struct gk_fib *
+init_grantor_fib_locked(const char *ip_prefix,
+	const char *grantor, struct gk_config *gk_conf)
+{
+	int ret, gt_fib_present, gt_prefix_len, ip_prefix_len;
+	struct ipaddr ip_prefix_addr, gt_addr;
+	struct gk_fib *gt_prefix_fib = NULL;
+
+	/* Initialize the fib entry for the Grantor. */
+	ret = convert_str_to_ip(grantor, &gt_addr);
+	if (ret < 0)
+		return NULL;
+
+	ip_prefix_len = parse_ip_prefix(ip_prefix, &ip_prefix_addr);
+	if (ip_prefix_len < 0)
+		return NULL;
+
+	if (gt_addr.proto == ETHER_TYPE_IPv4)
+		gt_prefix_len = sizeof(struct in_addr) * 8;
+	else
+		gt_prefix_len = sizeof(struct in6_addr) * 8;
+
+	/*
+	 * Find the FIB entry for the Grantor IP prefix.
+	 *
+	 * There must be an entry for it:
+	 * (1) The FIB entry for the Grantor exists.
+	 * (2) A FIB entry for the Grantor prefix.
+	 * (3) The Grantor is a neighbor, so it has a neighbor FIB entry.
+	 */
+	gt_prefix_fib = find_fib_entry_for_grantor_locked(
+		&gt_addr, gt_prefix_len, &gt_fib_present, gk_conf);
+	if (gt_prefix_fib == NULL)
+		return NULL;
+
+	/* It's the FIB entry for the Grantor. */
+	if (gt_fib_present)
+		return gt_prefix_fib;
+
+	if ((gt_prefix_fib->action != GK_FWD_GATEWAY_BACK_NET &&
+			gt_prefix_fib->action != GK_FWD_NEIGHBOR_BACK_NET)) {
+		return NULL;
+	}
+
+	return fill_up_fib_entry_for_grantor_locked(
+		&gt_addr, gt_prefix_len, gt_prefix_fib,
+		&ip_prefix_addr, ip_prefix_len, gk_conf);
+}
+
+static struct gk_fib *
+init_drop_fib_locked(const char *ip_prefix, struct gk_config *gk_conf)
+{
+	int fib_id;
+	int ip_prefix_len;
+	struct ipaddr ip_prefix_addr;
+	struct gk_fib *ip_prefix_fib;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	/* Initialize the fib entry for the IP prefix. */
+	ip_prefix_len = parse_ip_prefix(ip_prefix, &ip_prefix_addr);
+	if (ip_prefix_len < 0)
+		return NULL;
+
+	fib_id = get_empty_fib_id(ip_prefix_addr.proto, gk_conf);
+	if (fib_id < 0)
+		return NULL;
+
+	if (ip_prefix_addr.proto == ETHER_TYPE_IPv4)
+		ip_prefix_fib = &ltbl->fib_tbl[fib_id];
+	else if (likely(ip_prefix_addr.proto == ETHER_TYPE_IPv6))
+		ip_prefix_fib = &ltbl->fib_tbl6[fib_id];
+	else
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: unknown IP type %hu at %s",
+			ip_prefix_addr.proto, __func__);
+
+	ip_prefix_fib->action = GK_DROP;
+
+	return ip_prefix_fib;
+}
+
+static int
+add_fib_entry_locked(const char *ip_prefix, enum gk_fib_action action,
+	const char *grantor_or_gateway, struct gk_config *gk_conf)
+{
+	if (ip_prefix == NULL || grantor_or_gateway == NULL)
```

Comment:
If `grantor_or_gateway` cannot be `NULL`, how can one add a `GK_DROP` FIB entry?